### PR TITLE
Add unit testing

### LINKED ISSUE #35: Missing Unit Tests
Similar to #33, the code on a whole lacks unit testing. We should implement these tests sooner rather than later, since unit tests should ideally be created at around the same time the code that they're testing is made.

For the Rust portions of the project, I think the files are small enough that we can go with the standard Rustic way of adding unit testing in a module at the bottom of a file. Once the files get bigger, we can consider moving them off into their own folder. We will need to teach `colcon-cargo` how to run these tests, but that shouldn't be too difficult.

The other languages should follow the standard ROS 2 rules for unit testing.

### PR DESCRIPTION
**Pull request by esteve**
Closes #35

Supersedes #52

### CODE REVIEW

**Comment by mxgrey** [comment]
Adding unit tests to `ros2_rust` is awesome!

I tried running the tests on my machine, and they almost all pass except for this:

```
---- tests::test_spin_once stdout ----
Error: ReturnCode(Timeout)
thread 'tests::test_spin_once' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', /rustc/17d29dcdce9b9e838635eb0adefd9b8b1588410b/library/test/src/lib.rs:186:5
```

I'm extremely confused about where this assertion is happening. The backtrace mentions line 186, but there's no line 186 in `lib.rs`. Is anyone else getting this test failure, or something messed up in my environment?

**Comment by AlexKaravaev** [comment]
@mxgrey I have something similar. 

```test_spin_once``` in my case runs indefinitely. Output to the console is:

```
test tests::test_spin_once has been running for over 60 seconds
``` 

This is output of corresponding thread from gdb, not sure if it will help, but anyway pasting it here:

```
Thread 147 (Thread 0x7ffff6880700 (LWP 84856)):
#0  __lll_lock_wait (futex=futex@entry=0x7fffec001948, private=0) at lowlevellock.c:52
#1  0x00007ffff7cb80a3 in __GI___pthread_mutex_lock (mutex=0x7fffec001948) at ../nptl/pthread_mutex_lock.c:80
#2  0x00007ffff6fe5479 in ?? () from /opt/ros/galactic/lib/librmw_cyclonedds_cpp.so
#3  0x00007ffff6fe6ee8 in rmw_create_node () from /opt/ros/galactic/lib/librmw_cyclonedds_cpp.so
#4  0x00007ffff7d9e2e4 in rcl_node_init () from /opt/ros/galactic/lib/librcl.so
#5  0x0000555555596416 in rclrs::node::Node::new_with_namespace (node_name="minimal_subscriber", node_ns="", context=0x7ffff687ee10) at src/node/mod.rs:67
#6  0x000055555559616f in rclrs::node::Node::new (node_name="minimal_subscriber", context=0x7ffff687ee10) at src/node/mod.rs:51
#7  0x000055555559d15a in rclrs::context::Context::create_node (self=0x7ffff687ee10, node_name="minimal_subscriber") at src/context.rs:78
#8  0x00005555555a26d3 in rclrs::tests::test_spin_once () at src/lib.rs:135
#9  0x00005555555966ce in rclrs::tests::test_spin_once::{{closure}} () at src/lib.rs:130
#10 0x0000555555574e7e in core::ops::function::FnOnce::call_once () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/core/src/ops/function.rs:227
#11 0x00005555555ceb33 in core::ops::function::FnOnce::call_once () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/core/src/ops/function.rs:227
#12 test::__rust_begin_short_backtrace () at library/test/src/lib.rs:575
#13 0x00005555555cd7a7 in <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/alloc/src/boxed.rs:1854
#14 <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/core/src/panic/unwind_safe.rs:271
#15 std::panicking::try::do_call () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/std/src/panicking.rs:485
#16 std::panicking::try () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/std/src/panicking.rs:449
#17 std::panic::catch_unwind () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/std/src/panic.rs:136
#18 test::run_test_in_process () at library/test/src/lib.rs:598
#19 test::run_test::run_test_inner::{{closure}} () at library/test/src/lib.rs:492
#20 0x00005555555d6b01 in test::run_test::run_test_inner::{{closure}} () at library/test/src/lib.rs:519
#21 std::sys_common::backtrace::__rust_begin_short_backtrace () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/std/src/sys_common/backtrace.rs:123
#22 0x00005555555a374f in std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}} () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/std/src/thread/mod.rs:498
#23 <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/core/src/panic/unwind_safe.rs:271
#24 std::panicking::try::do_call () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/std/src/panicking.rs:485
#25 std::panicking::try () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/std/src/panicking.rs:449
#26 std::panic::catch_unwind () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/std/src/panic.rs:136
#27 std::thread::Builder::spawn_unchecked_::{{closure}} () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/std/src/thread/mod.rs:497
#28 core::ops::function::FnOnce::call_once{{vtable-shim}} () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/core/src/ops/function.rs:227
#29 0x000055555561e373 in <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/alloc/src/boxed.rs:1854
#30 <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/alloc/src/boxed.rs:1854
#31 std::sys::unix::thread::Thread::new::thread_start () at library/std/src/sys/unix/thread.rs:108
#32 0x00007ffff7cb5609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#33 0x00007ffff7a83163 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

```

**Comment by AlexKaravaev** [comment]
And switching DDS from CycloneDDS to FastRTPS results in the same result as yours

```
failures:

---- tests::test_spin_once stdout ----
(Got 0 messages so far)
Error: ReturnCode(Timeout)
thread 'tests::test_spin_once' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', /rustc/27f5d830eb11cd7bdc834d6f0d78120976f75443/library/test/src/lib.rs:187:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by nnmm** [comment]
FWIW, I think some of these can be implemented more elegantly as doctests – that way they serve a dual purpose. I'm adding some doctests to Context and Node currently.

**Comment by esteve** [comment]
Thanks @jhdcs and @Soya-Onishi for the original PR and the updates, respectively.

File: rclrs/src/node.rs [hunk 1]
```diff
@@ -286,3 +286,38 @@ unsafe fn call_string_getter_with_handle(
...
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{Context, Node, QOS_PROFILE_DEFAULT};
+
+    #[test]
+    fn test_new() -> Result<(), RclrsError> {
+        let context = Context::new(vec![]).expect("Context instatiation is expected to be success");
```
Comment (jhdcs) [inline rclrs/src/node.rs:297]:
```suggestion
        let context = Context::new(vec![]).expect("Context instantiation is expected to be a success");
```

File: rclrs/src/node.rs [hunk 2]
```diff
@@ -286,3 +286,38 @@ unsafe fn call_string_getter_with_handle(
...
+            "topic",
+            QOS_PROFILE_DEFAULT,
+            move |_: std_msgs::msg::String| {},
+        )?;
+        Ok(())
+    }
+
+    #[test]
+    fn test_create_publisher() -> Result<(), RclrsError> {
+        let context = Context::new(vec![]).expect("Context instatiation is expected to be success");
```
Comment (jhdcs) [inline rclrs/src/node.rs:317]:
```suggestion
        let context = Context::new(vec![]).expect("Context instantiation is expected to be a success");
```

File: rclrs/src/node/publisher.rs [hunk 3]
```diff
@@ -163,3 +163,38 @@ impl<'a, T: Message> MessageCow<'a, T> for &'a T {
...
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{Context, Publisher, QOS_PROFILE_DEFAULT};
+
+    fn create_fixture(name: &str) -> (Context, Node) {
+        let context =
+            Context::new(vec![]).expect("Context instantiation is expected to be success");
```
Comment (jhdcs) [inline rclrs/src/node/publisher.rs:174]:
```suggestion
            Context::new(vec![]).expect("Context instantiation is expected to be a success");
```

File: rclrs/src/node/publisher.rs [hunk 4]
```diff
@@ -163,3 +163,38 @@ impl<'a, T: Message> MessageCow<'a, T> for &'a T {
...
+        let node = context
+            .create_node(name)
+            .expect("Node instantiation is expected to be success");
```
Comment (jhdcs) [inline rclrs/src/node/publisher.rs:177]:
```suggestion
            .expect("Node instantiation is expected to be a success");
```

File: rclrs/src/node/subscription.rs [hunk 5]
```diff
@@ -206,3 +206,24 @@ where
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{Context, Subscription, QOS_PROFILE_DEFAULT};
+
+    #[test]
+    fn test_instantiate_subscriber() -> Result<(), RclrsError> {
+        let context =
+            Context::new(vec![]).expect("Context instantiation is expected to be success");
```
Comment (jhdcs) [inline rclrs/src/node/subscription.rs:218]:
```suggestion
            Context::new(vec![]).expect("Context instantiation is expected to be a success");
```

**Review by jhdcs — CHANGES_REQUESTED** [review]
Just some spelling errors. Otherwise looks good!

**Comment by esteve** [comment]
@jhdcs thanks for the review!

**Review by jhdcs — APPROVED** [review]

**Comment by esteve** [comment]
@Soya-Onishi I've merged this PR, but if you have any feedback, please add it here and I'll address it in a future PR

**Comment by Soya-Onishi** [comment]
@esteve Thanks for your notice!

**Pull request merged**